### ISSUE TITLE
`make install` fails due to missing curl

### ISSUE THREAD

**Issue by professor** [greenplum-db/gpdb#3168]
On a fresh macOS Sierra install, `make install` fails since curl is missing.

Looks like we need to modify config and add `brew install curl` to the macOS directions.

**Comment by hlinnaka** [comment]
That's odd, there is an autoconf check for libcurl, so it should fail at `./configure` already...

**Comment by hlinnaka** [comment]
A wild guess: Are you trying to build with `--without-libcurl`, but there is some code in the tree that uses libcurl without the proper #ifdefs around it? Perhaps in the new PXF code? What's the error you're seeing?

**Comment by professor** [comment]
I'm using `./configure --with-perl --with-python --with-libxml --prefix=/usr/local/gpdb` as per the readme.

**Comment by Eulerizeit** [comment]
What are the configure defaults? Does it default to --without-libcurl?

-- Rob

**Comment by professor** [comment]
I confirm that configure looks for curl....

```
checking for xmlSaveToBuffer in -lxml2... yes
checking for curl-config... /usr/bin/curl-config
checking for curl >= 7.19.0... yes
checking CURL_CFLAGS...
checking CURL_LIBS... -lcurl
checking whether CURLOPT_MAIL_FROM is declared... yes
checking for library containing BZ2_bzDecompress... -lbz2
...
configure: using LIBS=-ljson-c -lgpopt -lnaucrates -lgpdbcost -lgpos -lxerces-c -lbz2 -lxml2 -lz -ledit -lm  -lcurl
```
which curl
/usr/bin/curl

brew places curl at /usr/local/opt/curl/bin/curl

So what has changed in the last month that causes `make install` to fail? Hmmm.....

**Comment by hlinnaka** [comment]
Oh, that's coming from PXF's Makefile. It seems to do its own checks on curl. I cannot figure out why. @denalex ?

**Comment by professor** [comment]
@hlinnaka, how should extensions pull in system libraries that autoconf verified?

**Comment by hlinnaka** [comment]
You need to have:
```
SHLIB_LINK += -lcurl
```
In the Makefile, so that the extension is linked with it. That's it.

**Comment by lavjain** [comment]
@shivzone just merged https://github.com/greenplum-db/gpdb/pull/3167

@professor Could you please try again?

**Comment by professor** [comment]
Looks like https://github.com/greenplum-db/gpdb/pull/3167 did not fix the issues with pxf/Makefile. 

I suspect that if someone enables PXF, they would have the same issue with curl.

**Comment by denalex** [comment]
Here's the detailed reason for failing build on MacOS (only) -- PXF requires a higher version of curl than what MacOS is shipped with. "brew install curl" will bring updated curl, but will NOT replace the system libraries, but put the new ones in the Cellar and symlink from /usr/local/opt/curl directory. We do not want developers to update system curl on their machines either, so during make install phase of PXF we will copy shared libs from the curl distro installed by brew into gpdb deployment directory, replicating what a GPDB installer would do.

The simplest check was to see if the user has ran "brew install curl", that's why we are checking the /usr/local/opt/curl location.

Now, PXF is not enabled by default, so most developers would not see the issue. Whoever needs to build with PXF using --enable-pxf flag, will have to do "brew install curl". I also added this command to the MacOS env script: https://github.com/greenplum-db/gpdb/commit/4722802f54e680be082889f8e52ac4959560ac3a

**Comment by hlinnaka** [comment]
On 09/07/2017 01:35 AM, Alexander Denissov wrote:
> Here's the detailed reason for failing build on MacOS (only) -- PXF requires a higher version of curl than what MacOS is shipped with. "brew install curl" will bring updated curl, but will NOT replace the system libraries, but put the new ones in the Cellar and symlink from /usr/local/opt/curl directory. We do not want developers to update system curl on their machines either, so during make install phase of PXF we will copy shared libs from the curl distro installed by brew into gpdb deployment directory, replicating what a GPDB installer would do.
> 
> The simplest check was to see if the user has ran "brew install curl", that's why we are checking the /usr/local/opt/curl location.

Ok. Firstly, we really shouldn't be doing that kind of stuff in a 
Makefile. Leave it to the autoconf scripts to figure out where to find 
the dependencies. If you need a particular version of libcurl, modify 
the autoconf rule to check for it. Secondly, the backend itself uses 
libcurl, too. Things will probably get quite confused, if the backend 
links with a different version of libcurl than what PXF tries to link 
with. Thirdly, comments in the Makefile would've be nice.

What version of libcurl does PXF need? The backend requires version 
7.19.0, so that's what the autoconf rule checks currently. Does PXF need 
something newer? What feature from the newer libcurl do you need?

- Heikki

**Comment by professor** [comment]
Heikki, thanks for expressing so well questions that I was wondering about. I was also curious how an extension should describe dependencies. We've started using conan for expressing dependencies in other parts of the system, would that apply here as well?

**Comment by hlinnaka** [comment]
No idea, I'm not familiar with Conan.

**Comment by hlinnaka** [comment]
That's an orthogonal issue, though. Whether or not we use Conan to get the dependent libraries, the autoconf tests still need to be kept up-to-date. So the question remains, what version of libcurl does PXF reuire? What's the "critical" feature that determines the version required?

**Comment by danielgustafsson** [comment]
This is addressed in #3209.

**Comment by professor** [comment]
@denalex expressed some concern that this issue has morphed from the original one. I'm closing this issue and opening a new one here: https://github.com/greenplum-db/gpdb/issues/3214
